File: samples/python/17.multilingual-bot/translation/translation_middleware.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

from typing import Callable, Awaitable, List

from botbuilder.core import Middleware, UserState, TurnContext
from botbuilder.schema import Activity, ActivityTypes

from translation import MicrosoftTranslator
from translation.translation_settings import TranslationSettings


class TranslationMiddleware(Middleware):
    """
    Middleware for translating text between the user and bot.
    Uses the Microsoft Translator Text API.
    """

    def __init__(self, translator: MicrosoftTranslator, user_state: UserState):
        self.translator = translator
        self.language_preference_accessor = user_state.create_property(
            "LanguagePreference"
        )

    async def on_turn(
        self, context: TurnContext, logic: Callable[[TurnContext], Awaitable]
    ):
        """
        Processes an incoming activity.
        :param context:
        :param logic:
        :return:
        """
        translate = await self._should_translate(context)
        if translate and context.activity.type == ActivityTypes.message:
            context.activity.text = await self.translator.translate(
                context.activity.text, TranslationSettings.default_language.value
            )

        async def aux_on_send(
            ctx: TurnContext, activities: List[Activity], next_send: Callable
        ):
            user_language = await self.language_preference_accessor.get(
                ctx, TranslationSettings.default_language.value
            )
            should_translate = (
                user_language != TranslationSettings.default_language.value
            )

            # Translate messages sent to the user to user language
            if should_translate:
                for activity in activities:
                    await self._translate_message_activity(activity, user_language)

            return await next_send()

        async def aux_on_update(
            ctx: TurnContext, activity: Activity, next_update: Callable
        ):
            user_language = await self.language_preference_accessor.get(
                ctx, TranslationSettings.default_language.value
            )
            should_translate = (
                user_language != TranslationSettings.default_language.value
            )

            # Translate messages sent to the user to user language
            if should_translate and activity.type == ActivityTypes.message:
                await self._translate_message_activity(activity, user_language)

            return await next_update()

        context.on_send_activities(aux_on_send)
        context.on_update_activity(aux_on_update)

        await logic()

    async def _should_translate(self, turn_context: TurnContext) -> bool:
        user_language = await self.language_preference_accessor.get(
            turn_context, TranslationSettings.default_language.value
        )
        return user_language != TranslationSettings.default_language.value

    async def _translate_message_activity(self, activity: Activity, target_locale: str):
        if activity.type == ActivityTypes.message:
            activity.text = await self.translator.translate(
                activity.text, target_locale
            )

File: samples/java_springboot/17.multilingual-bot/src/main/java/com/microsoft/bot/sample/multilingual/translation/TranslationMiddleware.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.bot.sample.multilingual.translation;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import com.google.common.base.Strings;
import com.microsoft.bot.builder.Middleware;
import com.microsoft.bot.builder.NextDelegate;
import com.microsoft.bot.builder.StatePropertyAccessor;
import com.microsoft.bot.builder.TurnContext;
import com.microsoft.bot.builder.UserState;
import com.microsoft.bot.schema.Activity;
import com.microsoft.bot.schema.ActivityTypes;

/**
 * Middleware for translating text between the user and bot.
 * Uses the Microsoft Translator Text API.
 */
public class TranslationMiddleware implements Middleware  {
    private MicrosoftTranslator translator;
    private StatePropertyAccessor<String> languageStateProperty;

    /**
     * Initializes a new instance of the {@link TranslationMiddleware} class.
     * @param withTranslator Translator implementation to be used for text translation.
     * @param userState State property for current language.
     */
    public TranslationMiddleware(MicrosoftTranslator withTranslator, UserState userState) {
        if (withTranslator == null) {
            throw new IllegalArgumentException("withTranslator");
        }
        this.translator = withTranslator;
        if (userState == null) {
            throw new IllegalArgumentException("userState");
        }

        this.languageStateProperty = userState.createProperty("LanguagePreference");
    }

    /**
     * Processes an incoming activity.
     * @param turnContext Context object containing information for a single turn of conversation with a user.
     * @param next The delegate to call to continue the bot middleware pipeline.
     * @return A Task representing the asynchronous operation.
     */
    public CompletableFuture<Void> onTurn(TurnContext turnContext, NextDelegate next) {
        if (turnContext == null) {
            throw new IllegalArgumentException("turnContext");
        }

        return this.shouldTranslate(turnContext).thenCompose(translate -> {
            if (translate) {
                if (turnContext.getActivity().isType(ActivityTypes.MESSAGE)) {
                    return this.translator.translate(
                        turnContext.getActivity().getText(),
                        TranslationSettings.DEFAULT_LANGUAGE)
                    .thenApply(text -> {
                        turnContext.getActivity().setText(text);
                        return CompletableFuture.completedFuture(null);
                    });
                }
            }
            return CompletableFuture.completedFuture(null);
        }).thenCompose(task -> {
            turnContext.onSendActivities((newContext, activities, nextSend) -> {
                return this.languageStateProperty.get(turnContext, () -> TranslationSettings.DEFAULT_LANGUAGE).thenCompose(userLanguage -> {
                    Boolean shouldTranslate = !userLanguage.equals(TranslationSettings.DEFAULT_LANGUAGE);

                    // Translate messages sent to the user to user language
                    if (shouldTranslate) {
                        ArrayList<CompletableFuture<Void>> tasks = new ArrayList<CompletableFuture<Void>>();
                        for (Activity activity : activities.stream().filter(a -> a.getType().equals(ActivityTypes.MESSAGE)).collect(Collectors.toList())) {
                            tasks.add(this.translateMessageActivity(activity, userLanguage));
                        }

                        if (!Arrays.asList(tasks).isEmpty()) {
                            CompletableFuture.allOf(tasks.toArray(new CompletableFuture[tasks.size()])).join();
                        }
                    }

                    return nextSend.get();
                });
            });

            turnContext.onUpdateActivity((newContext, activity, nextUpdate) -> {
                return this.languageStateProperty.get(turnContext, () -> TranslationSettings.DEFAULT_LANGUAGE).thenCompose(userLanguage -> {
                    Boolean shouldTranslate = !userLanguage.equals(TranslationSettings.DEFAULT_LANGUAGE);

                    // Translate messages sent to the user to user language
                    if (activity.getType().equals(ActivityTypes.MESSAGE)) {
                        if (shouldTranslate) {
                            this.translateMessageActivity(activity, userLanguage);
                        }
                    }

                    return nextUpdate.get();
                });
            });

            return next.next();
        });
    }

    private CompletableFuture<Void> translateMessageActivity(Activity activity, String targetLocale) {
        if (activity.getType().equals(ActivityTypes.MESSAGE)) {
            return this.translator.translate(activity.getText(), targetLocale).thenAccept(text -> {
                activity.setText(text);
            });
        }
        return CompletableFuture.completedFuture(null);
    }

    private CompletableFuture<Boolean> shouldTranslate(TurnContext turnContext) {
        return this.languageStateProperty.get(turnContext, () -> TranslationSettings.DEFAULT_LANGUAGE).thenApply(userLanguage -> {
            if (Strings.isNullOrEmpty(userLanguage)) {
                userLanguage = TranslationSettings.DEFAULT_LANGUAGE;
            }
            return !userLanguage.equals(TranslationSettings.DEFAULT_LANGUAGE);
        });
    }
}
